File: backend/src/main/java/ru/nsu/fit/militarysystem/store/entity/MilitaryMan.java
package ru.nsu.fit.militarysystem.store.entity;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.Type;

import javax.persistence.*;
import java.time.LocalDate;
import java.util.LinkedHashSet;
import java.util.Set;

@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
@JsonIgnoreProperties({ "hibernateLazyInitializer", "handler" })
@Entity
@Table(name = "military_men", indexes = {
        @Index(name = "military_men_identification_number_key", columnList = "identification_number", unique = true)
})
public class MilitaryMan implements BaseEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id", nullable = false)
    private Integer id;

    @Lob
    @Type(type = "org.hibernate.type.TextType")
    @Column(name = "second_name", nullable = false)
    private String secondName;

    @Lob
    @Type(type = "org.hibernate.type.TextType")
    @Column(name = "first_name", nullable = false)
    private String firstName;

    @Lob
    @Type(type = "org.hibernate.type.TextType")
    @Column(name = "patronymic")
    private String patronymic;

    @Column(name = "date_of_birth", nullable = false)
    private LocalDate dateOfBirth;

    @Column(name = "date_of_award", nullable = false)
    private LocalDate dateOfAward;

    @Lob
    @Type(type = "org.hibernate.type.TextType")
    @Column(name = "identification_number")
    private String identificationNumber;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "rank_id", nullable = false)
    private Rank rank;

    @ManyToMany
    @JoinTable(name = "military_men__military_specialties",
               joinColumns = @JoinColumn(name = "military_man_id"),
               inverseJoinColumns = @JoinColumn(name = "military_specialty_id"))
    private Set<MilitarySpecialty> militarySpecialties = new LinkedHashSet<>();

    @ManyToMany
    @JoinTable(name = "military_men__military_formations",
               joinColumns = @JoinColumn(name = "military_man_id"),
               inverseJoinColumns = @JoinColumn(name = "military_formation_id"))
    private Set<MilitaryFormation> militaryFormations = new LinkedHashSet<>();
}

File: frontend/MilitarySystem/Models/Entities/MilitaryMan.swift
//
//  MilitaryMan.swift
//  MilitarySystem
//
//  Created by Mackem Meya on 12.05.2022.
//

import Foundation

struct MilitaryMan: Entity {
    var id: UInt32 = 0
    
    var secondName: String = ""
    
    var firstName: String = ""
    
    var patronymic: String = ""
    
    var dateOfBirth: Date = Date()
    
    var identificationNumber: String = ""
    
    var rankName: String = ""
    
    var dateOfAward: Date = Date()
    
    enum Field: String {
        case id = "ID"
        case secondName = "Фамилия"
        case firstName = "Имя"
        case patronymic = "Отчество"
        case dateOfBirth = "Дата рождения"
        case identificationNumber = "Идентификационный номер"
        case rankName = "Звание"
        case dateOfAward = "Дата получения звания"
    }
    
    static func < (lhs: MilitaryMan, rhs: MilitaryMan) -> Bool {
        return lhs.id < rhs.id
    }
}
